ngraph/frontend/onnx/frontend/src/core/graph.cpp

```diff
@@ -326,28 +357,48 @@ void Subgraph::find_inputs_from_parent() {
     for (const auto& node_proto : m_model->get_graph().node()) {
         int input_index = 0;
         for (const auto& in_name : node_proto.input()) {
-            if (m_parent_graph_cache->contains(in_name)) {
-                const auto& from_parent_node = m_parent_graph_cache->get_node(in_name);
+            if (m_parent_graph->is_ng_node_in_cache(in_name)) {
+                const auto& from_parent_node = m_parent_graph->get_ng_node_from_cache(in_name);
                 // constants are skipped
                 if (!ngraph::is_type<ngraph::op::Constant>(from_parent_node.get_node_shared_ptr())) {
                     for (const auto& out_name : node_proto.output()) {
                         if (m_cache->contains(out_name)) {
-                            auto out_node_to_replace_input = m_cache->get_node(out_name);
-                            auto new_param =
-                                std::make_shared<ngraph::op::Parameter>(from_parent_node.get_element_type(),
-                                                                        from_parent_node.get_partial_shape());
-                            // replace input from parent scope with parameter
-                            out_node_to_replace_input.get_node()->input(input_index).replace_source_output(new_param);
-                            m_parameter_to_parent_node_map.insert({new_param, in_name});
-                            m_cache->emplace_node(in_name, new_param);
-                            m_parameters.push_back(new_param);
-                            m_inputs_from_parent.push_back(in_name);
+                            auto node_to_replace_input = m_cache->get_node(out_name);
+                            replace_input_from_parent_scope_with_parameter(
+                                in_name,
+                                from_parent_node,
+                                node_to_replace_input.get_node()->input(input_index));
                         }
                     }
                 }
             }
             ++input_index;
         }
+        // Nodes with subgraphs (like Loop or If) can have implicit inputs (so their subgraphs depend on nodes from
+        // parent) Those implicit inputs are not present in `node_proto.input()` list so to get them, we need to fetch
+        // node's nGraph representation and then we can match those inputs with parent nodes
+        for (const auto& out_name : node_proto.output()) {
+            if (m_cache->contains(out_name)) {
+                auto node_to_replace_input = m_cache->get_node(out_name).get_node();
+                if (!dynamic_cast<op::util::SubGraphOp*>(node_to_replace_input) &&
```

Comment:
Since `SubGraphOp` inherits from `MultiSubGraphOp` separate checking `SubGraphOp` seems to be redundant.